### DISCUSSION TITLE
Allow to specify the location to generate Route Module types

### DISCUSSION THREAD

**Discussion by khabubuphathu** [remix-run/react-router#12450] [Proposals]
#### Problem
Currently, the Type Safety CLI in React Router does not provide an option to specify the location where the Route Module types should be generated. This limitation can lead to inconvenience for developers who may be using a monorepo and have specific package for types that they would like to share with other packages. Limiting developers to rely of the no typesafe option of `ActionFunctionArgs` and `LoaderFunctionArgs`

#### Solution
Introduce a new option in the Type Safety CLI that allows developers to specify the directory where the Route Module types should be generated. This option can be added as a command-line argument or a configuration setting in Vite plugin

#### Implementation
1. **Command-Line Argument**: Add a new argument to the Type Safety CLI command, e.g., `--types-dir`, to specify the directory for generated types.
2. **Vite Plugin File**: Allow the directory to be specified in the project's configuration file, e.g., `vite.config.ts`, under a new property like `typesDir`.

#### Example Usage
```sh
# Using command-line argument
react-router typegen --types-dir ../../packages/types
```

```typescript
// vite.config.ts
export default defineConfig({
  plugins: [
    reactRouter({
      typesDir: "../../packages/types"
    })
  ],
});
```

**Comment by rintaromasuda** [comment]
Thanks for bringing this up @khabubuphathu ! This is also what my team needs in a slightly different context.

Our Remix v2 app serves as multiple web apps by leveraging the following directory structure and an environment variable that is consumed in vite.config.ts:

```
root
 |- src
      |- apps
            |- app A
            |- app B
            |- app C
            |- app D
```

This works perfectly fines with Remix v2.

But since React Router v7 generates route types in the fixed location, that is ./react-router directory in the root, we realized that we can't generate route types for all our apps as generating types for one app wipes out ./react-router.

Ideally, just like @khabubuphathu said above, we can specify where we want to generate route types by command line option, or a config option (in react-router.config.ts or vite.config.ts) so that we can support multiple app scenarios.

**Reply by pcattori** [reply]
Could you share a small reproduction of your setup in a GitHub repo?

Mostly I'm curious about how the repo is structured w.r.t. `package.json`, `tsconfig.json`, `vite.config.ts`, etc. Does each app have its own copy of each of these files? Is there only one version of each of these files at the root of the project?

**Reply by pcattori** [reply]
@khabubuphathu if you have a small repro of your setup, that would also be useful!

**Comment by rintaromasuda** [comment]
Wondering if someone can comment on this as this is pretty much a blocking issue for us to migrate to React Router v7 from Remix v2.

**Comment by khabubuphathu** [comment]
You can continue to use types like `LoaderFunctionArgs` and `ActionFunctionArgs` to get the arguments of a loader/function so this shouldn't really block you. This is just a nice to have for strongly typed arguments.

**Comment by rintaromasuda** [comment]
@khabubuphathu Yeah, I know. That said, due to this issue, actually build fails for our multi-app site because of the race condition of the generated route types (Our build runs in parallel for the multi apps.) This said, it is really helpful if I can disable the route type generation.
